compiler/noirc_evaluator/src/ssa/opt/inline_const_brillig_calls.rs

```diff
@@ -0,0 +1,220 @@
+            instruction::{Instruction, InstructionId, TerminatorInstruction},
+            value::{Value, ValueId},
+        },
+        optimize_ssa_after_inline_const_brillig_calls, Ssa, SsaBuilder,
+    },
+};
+
+impl Ssa {
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn inline_const_brillig_calls(mut self, inliner_aggressiveness: i64) -> Self {
+        let error_selector_to_type = &self.error_selector_to_type;
+
+        // Collect all brillig functions so that later we can find them when processing a call instruction
+        let mut brillig_functions = BTreeMap::<FunctionId, Function>::new();
+        for (func_id, func) in &self.functions {
+            if let RuntimeType::Brillig(..) = func.runtime() {
+                let cloned_function = Function::clone_with_id(*func_id, func);
+                brillig_functions.insert(*func_id, cloned_function);
+            };
+        }
+
+        // Keep track of which brillig functions we couldn't completely inline: we'll remove the ones we could.
+        let mut brillig_functions_we_could_not_inline = HashSet::new();
+
+        for func in self.functions.values_mut() {
+            func.inline_const_brillig_calls(
+                &brillig_functions,
+                &mut brillig_functions_we_could_not_inline,
+                inliner_aggressiveness,
+                error_selector_to_type,
+            );
+        }
+
+        // Remove the brillig functions that are no longer called
+        for func_id in brillig_functions.keys() {
+            // We never want to remove the main function (it could be brillig if `--force-brillig` was given)
+            if self.main_id == *func_id {
+                continue;
+            }
+
+            if brillig_functions_we_could_not_inline.contains(func_id) {
+                continue;
+            }
+
+            // We also don't want to remove entry points
+            if self.entry_point_to_generated_index.contains_key(func_id) {
+                continue;
+            }
+
+            self.functions.remove(func_id);
+        }
+
+        self
+    }
+}
+
+impl Function {
+    pub(crate) fn inline_const_brillig_calls(
+        &mut self,
+        brillig_functions: &BTreeMap<FunctionId, Function>,
+        brillig_functions_we_could_not_inline: &mut HashSet<FunctionId>,
+        inliner_aggressiveness: i64,
+        error_selector_to_type: &BTreeMap<ErrorSelector, Type>,
+    ) {
+        for block_id in self.reachable_blocks() {
+            for instruction_id in self.dfg[block_id].take_instructions() {
+                if !self.optimize_const_brillig_call(
+                    instruction_id,
+                    brillig_functions,
+                    brillig_functions_we_could_not_inline,
+                    inliner_aggressiveness,
+                    error_selector_to_type,
+                ) {
+                    self.dfg[block_id].instructions_mut().push(instruction_id);
+                }
```

Comment:
FWIW I find this code a bit unintuitive at first glance. The docs of `optimize_const_brilling_call` say that:
> If the optimization is successful, returns true. Otherwise returns false. The given instruction is not removed from the function.

The instruction is "not removed from the function", but at the same time `.take_instructions()` removed all instructions from the block, and looking at the `if` it shows that if it optimised it then we are not supposed to add it back... is it _is_ removed, but not _by_ the optimise call?

Then I wonder what it does, and I suppose to optimise it needs to replace the instruction with something, so I wonder if it will push to `self.dfg` itself. And if it does, then why can't it decide whether to push the instruction back as-is, or not?

I'm still looking at the implementation below to understand what's going on, so it might get clearer in a minute, just wanted to mention it.

Reply:
Yeah, I see there is a lot of bookkeeping to be done on `self.dfg` in the successful case, and it's an elegant way to put the instruction back if it didn't happen, and at least the "putting back" is visually close to the taking. It's a shame there is no easy way to return the successful case to say "do this instead to the DFG".